threads.js

```diff
@@ -575,7 +704,106 @@ ManagerInternal.prototype.createThread = function(descriptor) {
   return process;
 };
 
-},{"./child-thread":2,"./utils":8}],6:[function(require,module,exports){
+},{"./child-thread":2,"./utils":10}],7:[function(require,module,exports){
+
+'use strict';
+
+var utils = require('./utils');
+
+module.exports = ServiceStream;
+
+/**
+ * Writable Stream instance passed to the `service.stream` implementation
+ *
+ * @param {String} id Stream ID used to sync client and service streams
+ * @param {BroadcastChannel} channel Channel used to communicate with client
+ * @param {String} serviceId ID of the service so that we can `postMessage`
+ * @param {String} clientId ID of the client that should receive messages
+ */
+function ServiceStream(id, channel, serviceId, clientId) {
+  this._ = new PrivateServiceStream(this, id, channel, serviceId, clientId);
+}
+
+/**
+ * Services that allows clients to cancel the operation before it's complete
+ * should override the `stream.cancel` method.
+ *
+ * @param {*} [reason] Data sent from client about the cancellation
+ */
+ServiceStream.prototype.cancel = function(reason) {
+  var err = new TypeError('service should implement stream.cancel()');
+  return Promise.reject(err);
+};
+
+/**
+ * Signal to client that action was aborted during the process, this should be
+ * used as a way to communicate errors.
+ *
+ * @param {*} [data] Reason of failure
+ * @return Promise
+ */
+ServiceStream.prototype.abort = function(data) {
+  return this._.post('abort', 'aborted', data);
+};
+
+/**
+ * Sends a chunk of data to the client.
+ *
+ * @param {*} data Chunk of data to be sent to client.
+ * @return Promise
+ */
+ServiceStream.prototype.write = function(data) {
+  return this._.post('write', 'writable', data);
+};
+
+/**
+ * Closes the stream, signals that action was completed with success.
+ *
+ * @return Promise
+ */
+ServiceStream.prototype.close = function() {
+  // according to whatwg streams spec, WritableStream#close() don't send data
+  return this._.post('close', 'closed');
+};
+
+function PrivateServiceStream(target, id, channel, serviceId, clientId) {
+  this.target = target;
+  this.id = id;
+  this.channel = channel;
+  this.client = clientId;
+  this.state = 'writable';
+  this.message = new utils.Messages(this, serviceId, []);
+}
+
+PrivateServiceStream.prototype.validateState = function(actionName, state) {
```

Comment:
it returns a promise because `cancel`, `abort`, `close` and `write` should only really send data to the client if it's on a valid state and they should return rejected promises when they fail.. so it made sense to do `return this.validateState('cancel', 'canceled').then(...)` but I'll replace the logic since it's confusing.